File: internal/images/contexts/_shared/models/comms.py
"""Models for agent communication channels.

Pydantic v2 schemas for messages, channels, and related types.
"""

import re
import uuid
from datetime import datetime, timezone
from enum import Enum

from pydantic import BaseModel, Field, field_validator, model_validator


class ChannelType(str, Enum):
    TEAM = "team"
    DIRECT = "direct"
    SYSTEM = "system"


class ChannelState(str, Enum):
    ACTIVE = "active"
    ARCHIVED = "archived"
    PURGED = "purged"


class MessageFlags(BaseModel):
    decision: bool = False
    question: bool = False
    blocker: bool = False
    urgent: bool = False
    approval_request: bool = False
    approval_response: bool = False


class Message(BaseModel):
    id: str = Field(default_factory=lambda: uuid.uuid4().hex[:12])
    channel: str
    author: str
    timestamp: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )
    content: str
    reply_to: str | None = None
    flags: MessageFlags = Field(default_factory=MessageFlags)
    metadata: dict = Field(default_factory=dict)
    edited_at: datetime | None = None
    edit_history: list[dict] = Field(default_factory=list)  # [{content: str, edited_at: str}]
    deleted: bool = False
    reactions: list[dict] = Field(default_factory=list)  # [{emoji: str, author: str}]

    @field_validator("content")
    @classmethod
    def content_bounded(cls, v: str) -> str:
        if len(v) > 10000:
            raise ValueError("Message content exceeds 10000 character limit")
        return v

    @model_validator(mode="after")
    def content_not_empty_unless_deleted(self) -> "Message":
        if not self.deleted and (not self.content or not self.content.strip()):
            raise ValueError("Message content cannot be empty")
        return self


_CHANNEL_NAME_RE = re.compile(r"^_?[a-z0-9][a-z0-9-]*[a-z0-9]$|^_?[a-z0-9]$")


class Channel(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    name: str
    type: ChannelType
    created_by: str
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )
    topic: str = ""
    members: list[str] = Field(default_factory=list)
    visibility: str = "open"
    state: ChannelState = ChannelState.ACTIVE
    deployment_id: str = ""
    base_name: str = ""  # original pack-defined channel name
    archived_at: datetime | None = None
    archived_by: str = ""

    @field_validator("visibility")
    @classmethod
    def validate_visibility(cls, v: str) -> str:
        if v not in ("open", "private", "platform-write"):
            raise ValueError(f"visibility must be 'open', 'private', or 'platform-write', got {v!r}")
        return v

    @field_validator("name")
    @classmethod
    def validate_channel_name(cls, v: str) -> str:
        if not _CHANNEL_NAME_RE.match(v):
            raise ValueError(
                f"Channel name must be lowercase kebab-case: {v!r}"
            )
        return v

File: internal/models/comms.go
// agency-gateway/internal/models/comms.go
package models

import (
	"fmt"
	"regexp"
	"strings"
	"time"
)

// Channel type constants.
const (
	ChannelTypeTeam   = "team"
	ChannelTypeDirect = "direct"
	ChannelTypeSystem = "system"
)

// DMChannelPrefix is the naming prefix for agent DM channels.
const DMChannelPrefix = "dm-"

// DMChannelName returns the DM channel name for an agent.
func DMChannelName(agentName string) string {
	return DMChannelPrefix + agentName
}

// IsDMChannel returns true if the channel name is a DM channel.
func IsDMChannel(name string) bool {
	return strings.HasPrefix(name, DMChannelPrefix)
}

// Channel state constants.
const (
	ChannelStateActive   = "active"
	ChannelStateArchived = "archived"
	ChannelStatePurged   = "purged"
)

// MessageFlags tracks message annotations.
type MessageFlags struct {
	Decision         bool `yaml:"decision" json:"decision"`
	Question         bool `yaml:"question" json:"question"`
	Blocker          bool `yaml:"blocker" json:"blocker"`
	Urgent           bool `yaml:"urgent" json:"urgent"`
	ApprovalRequest  bool `yaml:"approval_request" json:"approval_request"`
	ApprovalResponse bool `yaml:"approval_response" json:"approval_response"`
}

// Message represents a single channel message.
type Message struct {
	ID          string                   `yaml:"id" json:"id"`
	Channel     string                   `yaml:"channel" json:"channel" validate:"required"`
	Author      string                   `yaml:"author" json:"author" validate:"required"`
	Timestamp   time.Time                `yaml:"timestamp" json:"timestamp"`
	Content     string                   `yaml:"content" json:"content"`
	ReplyTo     *string                  `yaml:"reply_to" json:"reply_to,omitempty"`
	Flags       MessageFlags             `yaml:"flags" json:"flags"`
	Metadata    map[string]interface{}   `yaml:"metadata" json:"metadata"`
	EditedAt    *time.Time               `yaml:"edited_at" json:"edited_at,omitempty"`
	EditHistory []map[string]interface{} `yaml:"edit_history" json:"edit_history"`
	Deleted     bool                     `yaml:"deleted" json:"deleted"`
	Reactions   []map[string]interface{} `yaml:"reactions" json:"reactions"`
}

// Validate checks content bounds and non-empty when not deleted.
func (m *Message) Validate() error {
	if len(m.Content) > 10000 {
		return fmt.Errorf("message content exceeds 10000 character limit")
	}
	if !m.Deleted && strings.TrimSpace(m.Content) == "" {
		return fmt.Errorf("message content cannot be empty")
	}
	return nil
}

var channelNameRE = regexp.MustCompile(`^_?[a-z0-9][a-z0-9-]*[a-z0-9]$|^_?[a-z0-9]$`)

// Channel represents a communication channel.
type Channel struct {
	ID           string     `yaml:"id" json:"id"`
	Name         string     `yaml:"name" json:"name" validate:"required"`
	Type         string     `yaml:"type" json:"type" validate:"required,oneof=team direct system"`
	CreatedBy    string     `yaml:"created_by" json:"created_by" validate:"required"`
	CreatedAt    time.Time  `yaml:"created_at" json:"created_at"`
	Topic        string     `yaml:"topic" json:"topic"`
	Members      []string   `yaml:"members" json:"members"`
	Visibility   string     `yaml:"visibility" json:"visibility" validate:"required,oneof=open private platform-write" default:"open"`
	State        string     `yaml:"state" json:"state" validate:"required,oneof=active archived purged" default:"active"`
	DeploymentID string     `yaml:"deployment_id" json:"deployment_id"`
	BaseName     string     `yaml:"base_name" json:"base_name"`
	ArchivedAt   *time.Time `yaml:"archived_at" json:"archived_at,omitempty"`
	ArchivedBy   string     `yaml:"archived_by" json:"archived_by"`
}

// Validate checks channel name format.
func (c *Channel) Validate() error {
	if !channelNameRE.MatchString(c.Name) {
		return fmt.Errorf("channel name must be lowercase kebab-case: %q", c.Name)
	}
	return nil
}
